crates/ruma-client-api/src/profile.rs

```diff
@@ -1,7 +1,362 @@
+
+/// Trait implemented by types representing a field in a user's profile.
+#[cfg(feature = "unstable-msc4133")]
+pub trait ProfileField: Clone {
+    /// The type for the value of the field.
+    ///
+    /// This type should NOT be an `Option<_>`, the APIs already take and return optional values.
+    type Value: Sized + Serialize + DeserializeOwned;
+}
+
+/// Trait implemented by types representing a field in a user's profile having a statically-known
+/// name.
+#[cfg(feature = "unstable-msc4133")]
+pub trait StaticProfileField: ProfileField {
+    /// The `AnyProfileField` representation of this field.
+    fn as_profile_field() -> AnyProfileField;
+}
+
+/// Helper type to deserialize any type that implements [`StaticProfileField`].
+#[cfg(feature = "unstable-msc4133")]
+struct StaticProfileFieldVisitor<F: StaticProfileField>(PhantomData<F>);
+
+#[cfg(feature = "unstable-msc4133")]
+impl<'de, F: StaticProfileField> Visitor<'de> for StaticProfileFieldVisitor<F> {
+    type Value = <F as ProfileField>::Value;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(formatter, "a map with key `{}` and value", F::as_profile_field())
+    }
+
+    fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
+    where
+        V: MapAccess<'de>,
+    {
+        let field = F::as_profile_field();
+        let mut found = false;
+
+        while let Some(key) = map.next_key::<AnyProfileField>()? {
+            if key == field {
+                found = true;
+                break;
+            }
+        }
+
+        if !found {
+            return Err(de::Error::custom("missing field {field}"));
+        };
+
+        map.next_value()
+    }
+}
+
+/// The user's avatar URL.
+#[derive(Debug, Clone, Copy)]
+#[cfg(feature = "unstable-msc4133")]
+#[allow(clippy::exhaustive_structs)]
+pub struct AvatarUrl;
+
+#[cfg(feature = "unstable-msc4133")]
+impl ProfileField for AvatarUrl {
+    type Value = Option<OwnedMxcUri>;
+}
+
+#[cfg(feature = "unstable-msc4133")]
+impl StaticProfileField for AvatarUrl {
+    fn as_profile_field() -> AnyProfileField {
+        AnyProfileField::AvatarUrl
+    }
+}
+
+/// The user's display name.
+#[derive(Debug, Clone, Copy)]
+#[cfg(feature = "unstable-msc4133")]
+#[allow(clippy::exhaustive_structs)]
+pub struct DisplayName;
+
+#[cfg(feature = "unstable-msc4133")]
+impl ProfileField for DisplayName {
+    type Value = Option<String>;
+}
+
+#[cfg(feature = "unstable-msc4133")]
+impl StaticProfileField for DisplayName {
+    fn as_profile_field() -> AnyProfileField {
+        AnyProfileField::DisplayName
+    }
+}
+
+/// The possible fields of a user's profile.
+#[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/src/doc/string_enum.md"))]
+#[cfg(feature = "unstable-msc4133")]
+#[derive(Clone, PartialEq, Eq, PartialOrdAsRefStr, OrdAsRefStr, ruma_common::serde::StringEnum)]
+#[ruma_enum(rename_all = "snake_case")]
+#[non_exhaustive]
+pub enum AnyProfileField {
```

Comment:
This isn't very similar to the other `Any*` data-holding enums. How about we name this `ProfileFieldName`?

Reply:
Changed.